mirrord-layer/tests/fileops.rs

```diff
@@ -0,0 +1,266 @@
+use std::{collections::HashMap, path::PathBuf, process::Stdio, time::Duration};
+
+use actix_codec::Framed;
+use futures::{stream::StreamExt, SinkExt};
+use mirrord_protocol::{ClientMessage, DaemonCodec, DaemonMessage};
+use rstest::{fixture, rstest};
+use tokio::{
+    net::{TcpListener, TcpStream},
+    process::Command,
+};
+mod common;
+pub use common::*;
+
+struct LayerConnection {
+    codec: Framed<TcpStream, DaemonCodec>,
+}
+
+impl LayerConnection {
+    /// Accept a connection from the libraries and verify the first message it is supposed to send
+    /// to the agent - GetEnvVarsRequest. Send back a response.
+    /// Return the codec of the accepted stream.
+    async fn accept_library_connection(listener: &TcpListener) -> Framed<TcpStream, DaemonCodec> {
+        let (stream, _) = listener.accept().await.unwrap();
+        println!("Got connection from library.");
+        let mut codec = Framed::new(stream, DaemonCodec::new());
+        let msg = codec.next().await.unwrap().unwrap();
+        println!("Got first message from library.");
+        if let ClientMessage::GetEnvVarsRequest(request) = msg {
+            assert!(request.env_vars_filter.is_empty());
+            assert_eq!(request.env_vars_select.len(), 1);
+            assert!(request.env_vars_select.contains("*"));
+        } else {
+            panic!("unexpected request {:?}", msg)
+        }
+        codec
+            .send(DaemonMessage::GetEnvVarsResponse(Ok(HashMap::new())))
+            .await
+            .unwrap();
+        codec
+    }
+
+    /// Accept the library's connection and verify initial ENV message and PortSubscribe message
+    /// caused by the listen hook.
+    async fn get_initialized_connection(listener: &TcpListener) -> LayerConnection {
+        let codec = Self::accept_library_connection(listener).await;
+        LayerConnection { codec }
+    }
+
+    async fn is_ended(&mut self) -> bool {
+        self.codec.next().await.is_none()
+    }
+}
+
+/// Return the path to the existing layer lib, or build it first and return the path, according to
+/// whether the environment variable MIRRORD_TEST_USE_EXISTING_LIB is set.
+/// When testing locally the lib should be rebuilt on each run so that when developers make changes
+/// they don't have to also manually build the lib before running the tests.
+/// Building is slow on the CI though, so the CI can set the env var and use an artifact of an
+/// earlier job on the same run (there are no code changes in between).
+#[fixture]
+#[once]
+fn dylib_path() -> PathBuf {
```

Comment:
You're bringing `dylib_path` from `common` into scope but also redefining it here.